shared/state.rs

```diff
@@ -43,12 +43,18 @@ impl State {
     }
 
```

Comment:
Please add a few lines of comments. This function just became a lot more complex and deserves a little explanation :)